### ISSUE TITLE
successfully converted the model from temsorflow .pb to .mlmodel but not present in IOS

### ISSUE THREAD

**Issue by ChiFang** [tf-coreml/tf-coreml#126]
I successfully converted the model from temsorflow .pb to .mlmodel:
![image](https://user-images.githubusercontent.com/16358320/35978356-75f806b0-0d20-11e8-8745-ff643b64ef46.png)
but fail in prediction on IOS app:
<img width="1110" alt="error msg" src="https://user-images.githubusercontent.com/16358320/35978476-cb6cfd80-0d20-11e8-95c0-469472685813.png">
I have verified the log during convert. It seem OK:
![image](https://user-images.githubusercontent.com/16358320/35978513-ecf08102-0d20-11e8-8631-fa40a10eb2a9.png)
Thanks for any suggestion

**Comment by aseemw** [comment]
You are getting a validator error. This means the CoreML file produced by the converter is not valid. This is most likely due to a bug in the converter. 
I need to reproduce the  bug. Could you share the .pb file or the code block with the architecture of the model? 
If you do not want to share your model you can also file a bug at [Apple](https://developer.apple.com/bug-reporting/).

**Comment by ChiFang** [comment]
@aseemw  thanks a lot but how could I sent you my .pb?  e-mail?

**Comment by aseemw** [comment]
If you filing a bug report, you can attach the pb there. 
Otherwise, maybe you can upload it to your GitHub or upload it somewhere shareable.

**Comment by ChiFang** [comment]
@aseemw  here is my .pb and .py about how I convert
https://github.com/ChiFang/question/blob/master/coremlBug.rar

**Comment by ChiFang** [comment]
hi @aseemw  I think I found the root cause.
Once I replace  batch normalization from "tf.layers.batch_normalization" or "tf.contrib.slim_normalization" with "tf.nn.batch_normalization". IOS app won't return any error about .mlmodel.

**Comment by aseemw** [comment]
Okay. Good to know. However, the converter should work with the other two variants of BN as well. I'll keep this issue open and look into the .pb that you have attached.

**Comment by aseemw** [comment]
Hi @ChiFang The link you posted above is not working.

**Comment by ChiFang** [comment]
hi @aseemw  I remove it before because I think you douwnload already...sorry
I already upload again ^^

**Comment by aseemw** [comment]
I downloaded it, you can remove it if you want now!

**Comment by ChiFang** [comment]
Hi @aseemw I already convert my frozen .pb file to .mlmodel. (I have verified my .pb file predict result)
and use scripts in the utils/ to check .pb and .mlmodel

But my output by  .mlmodel (with Mac) is wrong compare to Tensorflow prediction by .pb file.
My input range is 0~1 without bias, so I call the convert just with:
tf_model_path, mlmodel_path, output_feature_names and input_name_shape_dict

my output range in Tensorflow .pb is about -20~+20 without error
the output range coreml .mlmodel is 10^n (n= 9~11).....   and give # different result even I input same image....... 

Thanks for any suggestion........

**Comment by aseemw** [comment]
Okay. Good to know that .mlmodel with coremltools gives correct predictions with TF. This means conversion was correct. 
Now for debugging your app results in Mac I have a few questions:

- How did you compare the predictions in python using coremltools. Did you feed in a numpy array or a PIL image to the .mlmodel? 
- What is the input type of your mlmodel? Is it an image or multi array? If its an image, then the input automatically will be in the range of [0,255]. You need to set the preprocessing parameters correctly so that image is scaled back in the range [0,1]

- What is your Mac software version
- Can you check if you see the deviations when you use the "usesCPUOnly" Flag set as true in your Mac app. ([API](https://developer.apple.com/documentation/coreml/mlpredictionoptions))
You would need to do something like this:
`MLPredictionOptions *options = [[MLPredictionOptions alloc] new];`
`options.usesCPUOnly = YES;`
`id<MLFeatureProvider> result = [_model.model predictionFromFeatures:modelInputObject options:options error:&err];`

**Comment by ChiFang** [comment]
wel... the .mlmodel with coremltools don't gives correct predictions. It just work without runtim evalidator error lol~~

my environment:
**training**: Ubuntu 16 with tensorflow 1.5 gpu version
**verify**: Mac with tensorflow 1.5 **cpu** version, last version coremltool and build from source tfcoreml
input: image read by opencv and scale range to  [0,1] by myself. (it's numpy array)
I have verified the range before call coreml predict on Mac with python by print Max and Min.

A: 
step1: I replace "batch_normalization" mentioned above and generated .pb file by [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py).

step2: convert .pb to .mlmodel by tfcoreml and check the models by the [utils](https://github.com/tf-coreml/tf-coreml/tree/master/utils)

step3: use .pb file to predict result by tensorflow in Mac and it's same from my ubuntu by .ckpt

step4: use .mlmodel file to predict result by coremltool in python and the result is totally wrong from .pb

Note: 
I do not verify .mlmodel in app but in python with coremltool predict because it is the fastest way to verify .mlmodel

**Comment by aseemw** [comment]
Can you update and run against PR #130. It should handle the model you had attached in https://github.com/ChiFang/question/blob/master/coremlBug.rar

**Comment by CR-Ko** [comment]
@aseemw , Thank you for your prompt assistance. This is very helpful.
